### PR TITLE
[FLINK-27162][runtime] Trigger non-periodic checkpoint in 'timer' thread

### PR DESCRIPTION
**Pull request by akalash**
## What is the purpose of the change

Checkpoint trigger always from one thread

## Brief change log
  - *Trigger non-periodic checkpoint in 'timer' thread *

## Verifying this change

 No extra tests

## Does this pull request potentially affect one of the following parts:

  - Dependencies (does it add or upgrade a dependency): (yes / **no**)
  - The public API, i.e., is any changed class annotated with `@Public(Evolving)`: (yes / **no**)
  - The serializers: (yes / **no** / don't know)
  - The runtime per-record code paths (performance sensitive): (yes / **no** / don't know)
  - Anything that affects deployment or recovery: JobManager (and its components), Checkpointing, Kubernetes/Yarn, ZooKeeper: (yes / **no** / don't know)
  - The S3 file system connector: (yes / **no** / don't know)

## Documentation

  - Does this pull request introduce a new feature? (yes / **no**)
  - If yes, how is the feature documented? (not applicable / docs / JavaDocs / not documented)

### CODE REVIEW

File: flink-runtime/src/main/java/org/apache/flink/runtime/checkpoint/CheckpointCoordinator.java [hunk 1]
```diff
@@ -74,11 +74,14 @@
...
 import java.util.concurrent.CompletableFuture;
 import java.util.concurrent.CompletionException;
+import java.util.concurrent.CountDownLatch;
 import java.util.concurrent.Executor;
 import java.util.concurrent.RejectedExecutionException;
 import java.util.concurrent.ScheduledFuture;
 import java.util.concurrent.ThreadLocalRandom;
 import java.util.concurrent.TimeUnit;
+import java.util.concurrent.atomic.AtomicBoolean;
+import java.util.concurrent.locks.LockSupport;
```
Comment (1996fanrui) [inline flink-runtime/src/main/java/org/apache/flink/runtime/checkpoint/CheckpointCoordinator.java:84]: Hi @akalash , thanks for your contribution. Could you run 'mvn spotless:apply' to remove these useless code?

**Comment by akalash** [comment]
I think we can rethink `triggerCheckpoint` and use it only from one `timer` thread(`main coordinator` thread in the future). Right now, I introduced the second method `triggerCheckpointFromCheckpointThread` because it is simple and it doesn't require a lot of changes but theoretically, we can use only one method `triggerCheckpoint` and if it is called from the non-expected thread(check the name of the thread?) then we can automatically wrap it with required thread(the same logic as we use in mailbox). Or at least we should check the condition that the `triggerCheckpoint` is called from the expected thread but it requires many changes in tests since many tests call this method directly from main test thread right now.

**Review by 1996fanrui — APPROVED** [review]
LGTM, thanks for the update.

File: flink-runtime/src/main/java/org/apache/flink/runtime/scheduler/adaptive/StateWithExecutionGraph.java [hunk 2]
```diff
@@ -286,7 +286,7 @@ CompletableFuture<String> triggerCheckpoint() {
         logger.info("Triggering a checkpoint for job {}.", jobID);
 
         return checkpointCoordinator
-                .triggerCheckpoint(false)
+                .triggerCheckpointFromCheckpointThread(false)
```
Comment (rkhachatryan) [inline flink-runtime/src/main/java/org/apache/flink/runtime/scheduler/adaptive/StateWithExecutionGraph.java:289]: The old method can now be made private. Then I'd named it `triggerCheckpointInternal` and name the new one (which uses `timer.execute`) `triggerCheckpoint`. 
The caller of the public method shouldn't chose which thread to use IMO.
WDYT?

**Review by rkhachatryan — COMMENTED** [review]
Thanks for the fix @akalash, in general it LGTM.
I've left a couple of comments regarding the interface though, PTAL.

File: flink-runtime/src/main/java/org/apache/flink/runtime/scheduler/adaptive/StateWithExecutionGraph.java (hunk 2 above)
Comment (akalash) [reply flink-runtime/src/main/java/org/apache/flink/runtime/scheduler/adaptive/StateWithExecutionGraph.java:289]: I thought about that. But I was afraid of the number of tests that use the synchronous triggering checkpoints. But now, as I see, perhaps, not all of them really require synchronous checkpoint. So I will take a look at how is difficult to change another tests.

File: flink-runtime/src/main/java/org/apache/flink/runtime/checkpoint/CheckpointCoordinator.java [hunk 3]
```diff
@@ -493,6 +498,15 @@ private CompletableFuture<CompletedCheckpoint> triggerSavepointInternal(
         return resultFuture;
     }
 
+    /**
+     * @param isPeriodic Flag indicating whether this triggered checkpoint is periodic.
+     * @return a future to the completed checkpoint.
+     */
+    public CompletableFuture<CompletedCheckpoint> triggerCheckpointFromCheckpointThread(
+            boolean isPeriodic) {
```
[the remark this answers is not in the document]
Comment (akalash) [reply flink-runtime/src/main/java/org/apache/flink/runtime/checkpoint/CheckpointCoordinator.java:506]: The same point about tests. But perhaps, I can migrate these tests to another method

**Comment by HuangXingBo** [comment]
Hi @akalash , could you rebase master and push again?

**Comment by HuangXingBo** [comment]
@flinkbot run azure

**Pull request closed without merging**